esp-hal/src/gpio/mod.rs

```diff
@@ -2249,60 +2384,110 @@ mod asynch {
         }
 
         /// Wait for the pin to undergo a transition from low to high.
+        ///
+        /// See [Self::wait_for] for more information.
         pub async fn wait_for_rising_edge(&mut self) {
             self.pin.wait_for_rising_edge().await
         }
 
         /// Wait for the pin to undergo a transition from high to low.
+        ///
+        /// See [Self::wait_for] for more information.
         pub async fn wait_for_falling_edge(&mut self) {
             self.pin.wait_for_falling_edge().await
         }
 
         /// Wait for the pin to undergo any transition, i.e low to high OR high
         /// to low.
+        ///
+        /// See [Self::wait_for] for more information.
         pub async fn wait_for_any_edge(&mut self) {
             self.pin.wait_for_any_edge().await
         }
     }
 
     #[must_use = "futures do nothing unless you `.await` or poll them"]
-    struct PinFuture {
-        pin_num: u8,
+    struct PinFuture<'d, P: InputPin> {
+        pin: Flex<'d, P>,
     }
 
-    impl PinFuture {
-        fn new(pin_num: u8) -> Self {
-            Self { pin_num }
+    impl<'d, P: InputPin> PinFuture<'d, P> {
+        fn pin_mask(&self) -> u32 {
+            let bank = self.pin.gpio_bank(private::Internal);
+            1 << (self.pin.number() - bank.offset())
+        }
+
+        fn is_done(&self) -> bool {
+            // Only the interrupt handler should clear the async bit, and only if the
+            // specific pin is handling an interrupt.
+            self.pin
+                .gpio_bank(private::Internal)
+                .async_operations()
+                .load(Ordering::Acquire)
+                & self.pin_mask()
+                == 0
         }
     }
 
-    impl core::future::Future for PinFuture {
+    impl<P: InputPin> core::future::Future for PinFuture<'_, P> {
         type Output = ();
 
         fn poll(self: core::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-            PIN_WAKERS[self.pin_num as usize].register(cx.waker());
+            PIN_WAKERS[self.pin.number() as usize].register(cx.waker());
 
-            // if pin is no longer listening its been triggered
-            // therefore the future has resolved
-            if !is_listening(self.pin_num) {
+            if self.is_done() {
                 Poll::Ready(())
             } else {
                 Poll::Pending
             }
         }
     }
+
+    impl<P: InputPin> Drop for PinFuture<'_, P> {
+        fn drop(&mut self) {
+            // If the pin isn't listening, the future has either been dropped before setup,
+            // or the interrupt has already been handled.
+            if self.pin.is_listening() {
+                // Make sure the future isn't dropped while the interrupt is being handled.
+                // This prevents tricky drop-and-relisten scenarios.
+                self.pin.unlisten();
+                while self.pin.is_interrupt_set() {}
```

Comment:
Is it possible that in between these two lines, the interrupt bit becomes set? So the interrupt handler doesn't run because `unlisten()` has been called but the while loop is stuck forever because the bit is set.

Reply:
How could an interrupt both set the status bit, but not actually end up generating an interrupt? We're not masking the CPU side, we're preventing the GPIO from generating the interrupt.

Reply:
GPIO looks a bit different so I may be wrong but typically you have an INT_RAW and a INT_ENA. INT_RAW will be set when the event happens and if INT_ENA is set, the handler is called.

`is_interrupt_set` looks at the INT_RAW and listen/unlisten modify INT_ENA.

What I'm trying to say here is, in between those two lines, the gpio event occurs (i.e. pin becomes high). `is_interrupt_set()` will start returning true but the handler won't run because `unlisten()` has been called.

Reply:
You are partly right in that the CPU interrupt itself looks like it can be masked by clearing INT_ENA. However, `unlisten` disables int_ena AND it writes 0 (disabled) to int_type. This means that if a GPIO event happens between the two lines you highlighted, nothing happens.

There is a different issue here, however:

- Event happens before unlisten, which is bound to be handled by the other core
- The other core is in a situation where it can't deal with the interrupt request (interrupt free context, higher priority handler, whatever).
- This core calls unlisten, which clears the interrupt request by clearing INT_ENA. (assuming that's how it works, this whole thing being undocumented is very weird and a BIG omission).
- The other core will not handle the interrupt request, because it is level triggered so the CPU forgets it.

I'm not immediately sure what we can do here - besides maybe never clearing INT_ENA once it's set. I think just setting INT_TYPE to 0 is equivalent in terms of sensitivity, but GPIO_STATUS will not get masked and the request eventually will get handled. Seems like an easy fix but I can't reason about its side effects, I've spent an hour researching this comment...

Reply:
Ahhh I see. I didn't understand what `unlisten()` was doing properly I see.

Cross-core async is hard to get right, which is why async drivers should just should just be kept `!Send`. Sadly it's not practical for GPIO with the way the API currently works.

Generally there should probably be some kind of ownership transfer between the application and the interrupt.
Perhaps a per-GPIO lock around `INT_ENA`/`INT_RAW`.
Users would be using `Mutex` around the gpio pins so they wouldn't actually have this problem if they did it themselves.

Reply:
Ooooo oh actually. Similar to the "is this pin async" bit flag you have, it can be a "is this pin async on this core".

So if an interrupt handler fires on a core, it can only handle gpio pins that listened on that core?